operator/builtin/input/http/operator.go

```diff
@@ -0,0 +1,216 @@
+	t.cancel = cancel
+	t.goListen(ctx)
+	return nil
+}
+
+// Stop will stop listening for log entries over http.
+func (t *HTTPInput) Stop() error {
+	t.cancel()
+	t.wg.Wait()
+	return nil
+}
+
+// goListenn will listen for http connections.
+func (t *HTTPInput) goListen(ctx context.Context) {
+	t.Debugf("using server config: %d", t.server.MaxHeaderBytes)
+
+	t.wg.Add(1)
+
+	entryCreateMethods := []string{"POST", "PUT"}
+
+	m := mux.NewRouter()
+	m.HandleFunc("/", t.goHandleMessages).Methods(entryCreateMethods...)
+
+	if t.auth != nil {
+		t.Debugf("using authentication middleware: %s", t.auth.name())
+		m.Use(t.auth.auth)
+	}
+
+	m.HandleFunc("/health", t.health).Methods("GET")
+
+	t.server.Handler = m
+
+	// shutdown go routine waits for a canceled context before stopping the server
+	go func() {
+		defer t.wg.Done()
+		for {
+			select {
+			case <-ctx.Done():
+				t.Debugf("Triggering http server shutdown")
+				ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+				defer cancel()
+				if err := t.server.Shutdown(ctx); err != nil {
+					t.Errorf("error while shutting down http server: %s", err)
+				}
+				return
+			default:
+				time.Sleep(time.Second * 2)
+			}
+		}
+	}()
+
+	// server go routine runs the http server
+	go func() {
+		t.Debugf("Starting http server on socket %s", t.server.Addr)
+		if err := t.server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			t.Errorf("http server failed: %s", err)
+			return
+		}
+		t.Debugf("Http server shutdown finished")
+	}()
+}
+
+// goHandleMessages will handles messages from a http connection by reading the request
+// body and returning http status codes.
+func (t *HTTPInput) goHandleMessages(w http.ResponseWriter, req *http.Request) {
+	ctx, cancel := context.WithCancel(req.Context())
+	t.wg.Add(1)
+	defer t.wg.Done()
+	defer cancel()
+
+	req.Body = http.MaxBytesReader(nil, req.Body, t.maxBodySize)
+	decoder := t.json.NewDecoder(req.Body)
+	body := make(map[string]interface{})
+	if err := decoder.Decode(&body); err != nil {
+		if strings.Contains(err.Error(), "too large") {
+			t.Errorf("failed to decode http %s request from %s: %s", req.Method, req.RemoteAddr, err)
+			w.WriteHeader(http.StatusRequestEntityTooLarge)
+			return
+		}
+		t.Errorf(err.Error())
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	entry, err := t.parse(body, req)
+	if err != nil {
+		t.Errorf("failed to create entry from http %s request from %s: %s", req.Method, req.RemoteAddr, err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	t.Write(ctx, entry)
+	w.WriteHeader(http.StatusCreated)
+}
+
+// parse will parse an http request's body into an entry
+func (t *HTTPInput) parse(body map[string]interface{}, req *http.Request) (*entry.Entry, error) {
+	if body == nil || req == nil {
+		return nil, fmt.Errorf("payload and http request must be set")
+	}
+
+	payload := make(map[string]interface{})
+
+	const msgKey = "message"
```

Comment:
What's the intent behind doing this additional layer of parsing and restructuring? I think the operator should accept any generic json payload and not mutate the received data. Instead, leave it up to the user to choose how to mutate their data in their pipeline setup.

If there's a strong argument to keep this, the behavior should be documented.

Reply:
The labels make a lot of sense though. I think we should add those. It's just the record alteration that I'm talking about.

Reply:
I agree. I tend to think too far ahead. The parsing will be handled by a plugin.